File: 0449-serialize-and-deserialize-bst/0449-serialize-and-deserialize-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string s;
        serializeHelper(root, s);
        return s;
    }
    
    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data.empty()) return nullptr;
        vector<int> nodes;
        stringstream ss(data);
        string token;
        while (ss >> token) {
            nodes.push_back(stoi(token));
        }
        int index = 0;
        return deserializeHelper(nodes, index, INT_MIN, INT_MAX);
    }
    
private:
    void serializeHelper(TreeNode* root, string& s) {
        if (!root) return;
        s += to_string(root->val) + " ";
        serializeHelper(root->left, s);
        serializeHelper(root->right, s);
    }
    
    TreeNode* deserializeHelper(vector<int>& nodes, int& index, int minVal, int maxVal) {
        if (index >= nodes.size()) return nullptr;
        int val = nodes[index];
        if (val < minVal || val > maxVal) return nullptr;
        
        index++;
        TreeNode* root = new TreeNode(val);
        root->left = deserializeHelper(nodes, index, minVal, val);
        root->right = deserializeHelper(nodes, index, val, maxVal);
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec* ser = new Codec();
// Codec* deser = new Codec();
// string tree = ser->serialize(root);
// TreeNode* ans = deser->deserialize(tree);
// return ans;

File: 0449-serialize-and-deserialize-bst/0449-serialize-and-deserialize-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
var serialize = function(root) {
    const res = [];
    serializeHelper(root, res);
    return res.join(' ');
};

function serializeHelper(root, res) {
    if (!root) return;
    res.push(root.val);
    serializeHelper(root.left, res);
    serializeHelper(root.right, res);
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
var deserialize = function(data) {
    if (!data || data.trim() === '') return null;
    const nodes = data.split(' ').map(Number);
    let index = 0;
    
    return deserializeHelper(nodes, () => {
        return nodes[index++];
    }, Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER);
};

function deserializeHelper(nodes, getNext, min, max) {
    if (index >= nodes.length) return null; // Use global index or pass by ref
    
    const val = getNext();
    if (val < min || val > max) {
        index--; // rollback
        return null;
    }
    
    const root = new TreeNode(val);
    root.left = deserializeHelper(nodes, getNext, min, val);
    root.right = deserializeHelper(nodes, getNext, val, max);
    return root;
}

// Fix: better index management
var deserialize = function(data) {
    if (!data) return null;
    const nodes = data.trim().split(/\s+/).map(Number);
    let i = 0;
    
    const build = (min, max) => {
        if (i >= nodes.length) return null;
        const val = nodes[i];
        if (val < min || val > max) return null;
        i++;
        const node = new TreeNode(val);
        node.left = build(min, val);
        node.right = build(val, max);
        return node;
    };
    
    return build(Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER);
};